File: June-LeetCoding-Challenge-2020/Week1/406-queue-reconstruction-by-height.js
/**
 * @param {number[][]} people
 * @return {number[][]}
 * Runtime: 92 ms, faster than 66.55% of JavaScript online submissions for Queue Reconstruction by Height.
 * Memory Usage: 41.5 MB, less than 23.33% of JavaScript online submissions for Queue Reconstruction by Height.
 * TC: O(N^2), where N is the length of people
 * SC: O(logN), where N is the length of people assuming we don't count return memory
 */
var reconstructQueue = function(people) {
    if (!people) {
        return people;
    }

    people.sort((a, b) => {
        if (a[0] !== b[0]) {
            return b[0] - a[0];
        } else {
            return a[1] - b[1];
        }
    });

    return people.reduce((acc, cur) => {
        acc.splice(cur[1], 0, cur);

        return acc;
    }, []);
};

File: June-LeetCoding-Challenge-2020/Week1/406-queue-reconstruction-by-height.py
class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        n = len(people)
        people.sort() # sort ppl by height and k in ascending order
        return_list = [None]*n
        for height, index in people:
            i,j = 0, -1 # i is the index of return_list
                        # j is the index of the queue with the greater than or equal height
            while(i<n):
                if not return_list[i] or return_list[i][0] == height:
                    j += 1
                if j == index:
                    break
                i+=1
            return_list[i] = [height, index]

        return return_list
        # TC: O(n^2)
        # SC: O(n), where n is the number of ppl in queue
        # Runtime: 1120 ms, faster than 6.32%
        # Memory Usage: 14.3 MB, less than 24.17%